File: WoofAndroid/presentation/profile/src/main/kotlin/com/perrystreet/woof/presentation/profile/viewmodel/ProfilePagerViewModel.kt
package com.perrystreet.woof.presentation.profile.viewmodel

import com.perrystreet.woof.presentation.common.rx.UiObservable
import com.perrystreet.woof.presentation.common.rx.UiObservable.Companion.asUiObservable
import com.perrystreet.woof.presentation.common.viewmodel.StateDerivingViewModel
import com.perrystreet.woof.presentation.profile.mapper.DogDomainToProfilePageUIModelMapper
import com.perrystreet.woof.presentation.profile.uimodel.ProfilePageUIModel
import com.perrystreet.woof.usecase.dogs.GetDogsFeedUseCase
import org.koin.android.annotation.KoinViewModel
import org.koin.core.annotation.InjectedParam

@KoinViewModel
class ProfilePagerViewModel(
    @InjectedParam private val dogId: Long,
    private val pageMapper: DogDomainToProfilePageUIModelMapper,
    getDogsFeedUseCase: GetDogsFeedUseCase,
) : StateDerivingViewModel<ProfilePagerViewModel.State>() {

    data class State(
        val pages: List<ProfilePageUIModel>,
        val initialPage: Int,
    ) {
        companion object {
            val Empty = State(pages = emptyList(), initialPage = 0)
        }
    }

    override val state: UiObservable<State> = getDogsFeedUseCase()
        .map { feed ->
            val pages = feed.dogs.map { pageMapper(it) }
            State(
                pages = pages,
                initialPage = pages.indexOfFirst { it.id == dogId }.coerceAtLeast(0),
            )
        }
        .distinctUntilChanged()
        .asUiObservable(State.Empty)
}

File: WoofIOS/SwiftPackages/Presentation/Profile/Sources/PresentationProfile/ViewModel/ProfilePagerViewModel.swift
import Combine
import DI
import Models
import PresentationCommon
import UseCase

@Factory
public final class ProfilePagerViewModel: StateDerivingViewModel<ProfilePagerViewModel.State, Never> {
    @DIArgument private let dogId: Int
    private let pageMapper: DogDomainToProfilePageUIModelMapper

    init(
        dogId: Int,
        pageMapper: DogDomainToProfilePageUIModelMapper,
        getDogsFeedUseCase: GetDogsFeedUseCase
    ) {
        self.dogId = dogId
        self.pageMapper = pageMapper
        super.init(
            initialValue: .empty,
            source: getDogsFeedUseCase()
                .map { feed in
                    let pages = feed.dogs.map { pageMapper(dog: $0) }
                    return State(
                        pages: pages,
                        initialPage: max(pages.firstIndex { $0.id == dogId } ?? 0, 0)
                    )
                }
                .removeDuplicates()
                .eraseToAnyPublisher()
        )
    }
}

extension ProfilePagerViewModel {
    public struct State: Equatable {
        public let pages: [ProfilePageUIModel]
        public let initialPage: Int

        public static let empty = State(pages: [], initialPage: 0)

        public init(pages: [ProfilePageUIModel], initialPage: Int) {
            self.pages = pages
            self.initialPage = initialPage
        }
    }
}
